content/en/blog/_posts/0000-00-00-kube-proxy-non-privileged.md

```diff
@@ -0,0 +1,107 @@
+---
+
+**Author**: Lars Ekman
+
+This post describes how the `--init-only` flag to `kube-proxy` can be
+used to run the main kube-proxy container in a stricter
+`securityContext` by performing the configuration that requires
+privileged mode in a separate [init container](
+/docs/concepts/workloads/pods/init-containers/). Since
+Windows doesn't have the equivalent of `capabilities`, this only works
+on Linux.
+
+The `kube-proxy` Pod still only meets the *privileged* [Pod Security
+Standard](https://kubernetes.io/docs/concepts/security/pod-security-standards/),but there is still an improvement because the running container doesn't
+need to run privileged.
+
+Please note that `kube-proxy` can be installed in different ways. The
+examples below assume that kube-proxy is run from a pod, but similar
+changes could be made in clusters where it is run as a system service.
+
+
+## Background
+
+It is undesirable to run a server container like `kube-proxy` in
+privileged mode. Security aware users wants to use capabilities instead.
+
+If `kube-proxy` is installed as a POD, the initialization requires
+"privileged" mode, mostly for setting sysctl's. However, `kube-proxy`
+only tries to set the sysctl's if they don't already have the right
+values. In theory, then, if a privileged [init container](/docs/concepts/workloads/pods/init-containers/)
+set the sysctls to the right values, then `kube-proxy` could run
+unprivileged.
+
+The problem is to know *what* to setup. Until now the only option has
+been to read the source to see what changes `kube-proxy` would have
+made, but with `--init-only` you can have `kube-proxy` itself do the setup
+*exactly* as on a normal start, and then exit.
+
+
+## Initializing kube-proxy in an init container
+
+Usually, cluster operators run `kube-proxy` in a privileged security context. Here's a snippet of an
+example manifest:
+
+```yaml
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  labels:
+    k8s-app: kube-proxy
+spec:
+  template:
+    spec:
+      containers:
+      - name: kube-proxy
+        command:
+        - /usr/local/bin/kube-proxy
+        - --config=/var/lib/kube-proxy/config.conf
+        - --hostname-override=$(NODE_NAME)
+        securityContext:
+          privileged: true
+```
+
+But now it is possible to use:
+
+```yaml
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  labels:
+    k8s-app: kube-proxy
+spec:
+  template:
+    spec:
+      initContainers:
+      - name: kube-proxy-init
+        command:
+        - /usr/local/bin/kube-proxy
+        - --config=/var/lib/kube-proxy/config.conf
+        - --hostname-override=$(NODE_NAME)
+        - --init-only
+        securityContext:
+          privileged: true
+      containers:
+      - name: kube-proxy
+        command:
+        - /usr/local/bin/kube-proxy
+        - --config=/var/lib/kube-proxy/config.conf
+        - --hostname-override=$(NODE_NAME)
+        securityContext:
+          capabilities:
+            add: ["NET_ADMIN"]
```

Comment:
Try this, which I think is a complete manifest.
```suggestion
apiVersion: apps/v1
kind: DaemonSet
metadata:
  namespace: kube-system
  name: kube-proxy-linux
  labels:
    k8s-app: kube-proxy
spec:
  selector:
    matchLabels:
      k8s-app: kube-proxy
      kubernetes.io/os: linux
  template:
    metadata:
      labels:
        k8s-app: kube-proxy
        kubernetes.io/os: linux
      annotations:
        kubernetes.io/description: >-
          Node-level network proxy; part of Kubernetes
    spec:
      tolerations:
      - key: node-role.kubernetes.io/control-plane
        effect: NoSchedule
        operator: Exists
      os: { name: linux } # On Windows, kube-proxy needs to run privileged
                        # You will need a second DaemonSet if your cluster
                        # includes Windows nodes
      nodeSelector:
        kubernetes.io/os: linux
      initContainers:
      - name: kube-proxy-init
        image: registry.k8s.io/kube-proxy:v1.28.2 # adjust this to the Kubernetes version you use
        imagePullPolicy: IfNotPresent
        # you should set resource requests here
        # see https://k8s.io/docs/concepts/configuration/manage-resources-containers/
        command:
        - /usr/local/bin/kube-proxy
        - --config=/var/lib/kube-proxy/config.conf
        - --hostname-override=$(NODE_NAME)
        - --init-only
        securityContext:
          privileged: true
      containers:
      - name: kube-proxy
        image: registry.k8s.io/kube-proxy:v1.28.2 # adjust this to the Kubernetes version you use
        imagePullPolicy: IfNotPresent
        # you should set resource requests here
        # see https://k8s.io/docs/concepts/configuration/manage-resources-containers/
        command:
        - /usr/local/bin/kube-proxy
        - --config=/var/lib/kube-proxy/config.conf
        - --hostname-override=$(NODE_NAME)
        securityContext:
          capabilities:
            add: ["NET_ADMIN"]
    terminationGracePeriodSeconds: 30
```

Reply:
I guess this doesn't cover a ConfigMap though, and maybe we should talk about that.

Reply:
I disagree on this. The kubeadm maintainers, and others who really will use this, are totally capable of figuring out the exact procedure. IMHO the blog should give normal users an _idea_ on what is done, not be a complete instruction. It should be TL;DR; Besides, the kubeadm manifest is likely not the only one used

Reply:
The confirMap is not affected. `--init-only` is a CLI option only, it can't be configured in the config file

Reply:
OK, but: every bit we miss out, let's clearly tell readers what's not there. We don't need to give all the details but let's make it **very very** obvious that this manifest is incomplete.

Even if we do all that, the website repo is likely to get issues filed by people who don't get it. Let's do what we reasonably can to minimize the rate of those issues.

Reply:
Updated.

It might be even more issues if we give the _illusion_ that the manifests are complete, and are compatible with the ones the users may have (which may not be kubeadm generated).

Reply:
So, on that: let's add a very clear note that this still isn't complete: nothing populates `/var/lib/kube-proxy/config.conf`.